src/Jackalope/Version/VersionHandler.php

```diff
@@ -0,0 +1,224 @@
+        $rootVersion->setProperty('jcr:successors', array()); // not part of the specification, but seems to be required
+        $additionalOperations[] = new AddNodeOperation($rootVersion->getPath(), $rootVersion);
+
+        // TODO add frozen node to root version
+
+        $node->setProperty('jcr:versionHistory', $versionHistory);
+        $node->setProperty('jcr:baseVersion', $rootVersion);
+        $node->setProperty('jcr:predecessors', array($rootVersion));
+
+        return $additionalOperations;
+    }
+
+    /**
+     * Performs a checkin for the node on the given path
+     *
+     * @param string $path The absolute path of the node to checkin
+     *
+     * @return string The path to the node containing the version information
+     */
+    public function checkinItem($path)
+    {
+        $node = $this->objectManager->getNodeByPath($path);
+
+        if (!$node->isNodeType(static::MIX_SIMPLE_VERSIONABLE)) {
+            throw new UnsupportedRepositoryOperationException(
+                'Node has to implement at least "mix:versionable" to use verisoning operations'
+            );
+        }
+
+        if ($node->isModified()) {
+            throw new InvalidItemStateException(
+                sprintf(
+                    'Node "%s" contains unsaved changes',
+                    $path
+                )
+            );
+        }
+
+        if (!$node->isCheckedOut()) {
+            return $path;
+        }
+
+        if ($node->hasProperty('jcr:mergeFailed')) {
+            throw new VersionException(sprintf('Node "%s" contains unresolved merge conflicts', $path));
+        }
+
+        // TODO set subgraph to read only
+
+        /** @var NodeInterface $versionHistoryNode */
+        $versionHistoryNode = $node->getPropertyValue('jcr:versionHistory');
+
+        // FIXME add some kind of sharding
+        /** @var NodeInterface $versionNode */
+        $versionNode = $versionHistoryNode->addNode(UUIDHelper::generateUUID(), 'nt:version');
+        $versionNode->setProperty('jcr:uuid', UUIDHelper::generateUUID());
+        $versionNode->setProperty('jcr:created', new \DateTime());
+        $versionNode->setProperty('jcr:successors', array());
+
+        // TODO create frozen node
+        $frozenNode = $versionNode->addNode('jcr:frozenNode', 'nt:frozenNode');
+        $frozenNode->setProperty('jcr:frozenPrimaryType', $node->getProperty('jcr:primaryType'));
+        if ($frozenNode->hasProperty('jcr:mixinTypes')) {
+            $frozenNode->setProperty('jcr:frozenMixinTypes', $node->getProperty('jcr:mixinTypes'));
+        }
+        $frozenNode->setProperty('jcr:frozenUuid', $node->getProperty('jcr:uuid'));
+
+        /** @var NodeInterface $baseVersionNode */
+        $baseVersionNode = $node->getPropertyValue('jcr:baseVersion');
+        $baseVersionNode->setProperty('jcr:successors', array($versionNode), PropertyType::REFERENCE, false);
+        $baseVersionNode->setModified();
```

Comment:
I find myself doing this quite often, since `setProperty` doesn't handle the state at all, if it is called with `$validate = false`. I am not sure if this is ok this way, or we should change the behavior of the `setProperty`-method resp. the subsequent function calls. /cc @dbu @lsmith77 @dantleech